lib/checks/aria/allowed-role.js

```diff
@@ -0,0 +1,207 @@
+		'doc-errata', 'doc-example', 'doc-foreword', 'doc-index', 'doc-introduction',
+		'doc-notice', 'doc-pagelist', 'doc-part', 'doc-preface', 'doc-prologue',
+		'doc-pullquote', 'doc-qna', 'doc-toc'
+	],
+	'svg': [ 'application', 'document', 'img' ],
+	'ul': [
+		'directory', 'group', 'listbox', 'menu', 'menubar', 'radiogroup',
+		'tablist', 'toolbar', 'tree', 'presentation'
+	],
+	'video': [ 'application' ]
+};
+
+/**
+ * Finds nearest parent with a specifig tagName
+ * @param  {HTMLElement} 		el      		child - starting pointer
+ * @param  {Array<String>} 	tagName 		Array containg capatilized tagnames
+ * @return {HTMLElement}         				Parent that matches one of the tagnames
+ */
+function getParentWithTagName(el, tagName) {
+	while (el.parentNode){
+		if(tagName.indexOf(el.tagName) > -1) { return el; }
+		el = el.parentNode;
+	}
+}
+
+/**
+ * Checks if given role is allowed for given tag
+ * @param  {string}  tagName key of allowedSubsets
+ * @param  {string}  role    current role
+ * @return {Boolean}         True if allowed
+ */
+function hasAllowedRole(tagName, role) {
+	return allowedSubsets[tagName].indexOf(role) > -1;
+}
+
+
+/**
+ * Contains a function for each htmlTag where not all roles allowed
+ * @type {Object}
+ */
+var customElementCheck = {
+	a: (el, role) => {
+		if(el.href) {
+			return hasAllowedRole('aWithHref', role);
+		} else {
+			return true;
+		}
+	},
+	area: (el) => !(el.href),
+	button: (el, role) => {
+		if(el.type === 'menu' ) {
+			return role === 'menuitem';
+		}
+		return hasAllowedRole('button', role);
+	},
+	img: (el, role) => {
+		var hasAllowedEmptyAltRole = hasAllowedRole('imgWithEmptyAlt', role);
+
+		if(el.alt) {
+      // any role exept the roles used by empty alt values
+			return hasAllowedEmptyAltRole ? 'img' : role;
+		} else {
+			return hasAllowedEmptyAltRole;
+		}
+	},
+	input: (el, role) => {
+		switch(el.type) {
+			case 'button':
+				return hasAllowedRole('inputTypeButton', role);
+			case 'checkbox':
+				return hasAllowedRole('inputTypeCheckbox', role);
+			case 'image':
+				return hasAllowedRole('inputTypeImage', role);
+			case 'radio':
+				return role === 'menuitemradio';
+			default:
+				return false;
+		}
+	},
+	li: (el, role) => {
+		let hasImplicitListitemRole = getParentWithTagName(el, ['OL', 'UL']);
+
+		if(hasImplicitListitemRole) {
+			return hasAllowedRole('li', role);
+		} else {
+			return true;
+		}
+	},
+	link: (el) => !(el.href),
+	menu: (el) => {
+		if(el.type === 'context'){
+			return false;
+		}
+		return true;
+	},
+	option: (el) => {
+		let withinOptionList = ['select', 'optgroup', 'datalist'].indexOf(el.parentNode) > -1;
+		return !withinOptionList;
+	},
+	select: (el, role) => (!el.multiple && el.size <= 1 && role === 'menu'),
+	td: (el) => !!getParentWithTagName(el, ['TABLE'])
+};
+
+var role = node.getAttribute('role');
+var tagName = node.tagName.toLowerCase();
+
+// check if tag is allowed to have a role
+if(elementsHasNoAllowedRoles.indexOf(tagName) > -1) {
```

Comment:
We should consider allowing users to whitelist certain elements through options, like we did with some of the other aria checks. Have a look here how this works:

https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/valid-attr.js#L11